stepanblyschak/sonic-utilities
# [techsupport] improve robustness

Signed-off-by: Stepan Blyschak <stepanb@nvidia.com>



#### What I did

1. Execute the command in a seperate bash process under timeout. Due to with '--foreground' the child processes are not killed and still have the file descriptor opened to piped process which will basically hang forever.
From the "timeout" man page:
```
       --foreground

              when not running timeout directly from a shell prompt,

              allow COMMAND to read from the TTY and get TTY signals; in
              this mode, children of COMMAND will not be timed out
```

So, a command like this:

```
timeout --foreground 5s some_process_that_spawn_childs_that_produce_output | tee some_file
```

Will actually hang forever. The '--foreground' option is still needed. So the proposed solution is to run that construct in another bash process:

```
timeout --foreground 5s bash -c "some_process_that_spawn_childs_that_produce_output | tee some_file"
```

2. Moved hw-mgmt-dump to collect_mellanox function and execute it under timeout.
3. Handle global timeout: When global timeout is reached the SIGTERM is sent to generate_dump process, upon SIGTERM the generate_dump process will finilize its work by creating a dump of what already has been collected, providing a dump even when timeout is reached. The generate_dump will be killed in case it cannot create dump after initial SIGTERM.

#### How I did it

I implemented the above mentioned features.

#### How to verify it

Artificially hang the FW, execute "show techsupport" and verify even in case timeout is reached the dump is generated.

#### Previous command output (if the output of a command-line utility has changed)

#### New command output (if the output of a command-line utility has changed)
